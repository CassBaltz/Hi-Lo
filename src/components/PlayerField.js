import React, { useState, useRef } from 'react';
import { withStyles } from '@material-ui/styles';
import { connect } from "react-redux";

import { updateName } from "../redux/actions/game";

const ENTER_KEY = 13;

const EditIcon = ({ classes }) => (
  <svg
    className={classes.editIcon}
    xmlns="http://www.w3.org/2000/svg"
    height="24"
    viewBox="0 0 24 24"
    width="24"
  >
    <path d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34c-.39-.39-1.02-.39-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z" />
    <path d="M0 0h24v24H0z" fill="none" />
  </svg>
);

export const PlayerField = ({ dispatch, classes, playerKey, name, }) => {
  const [isEditing, setIsEditing] = useState(false);
  const inputRef = useRef(null);

  const handleBlur = e => {
    if (!e.target.value) {
      return setIsEditing(true);
    }

    setIsEditing(false);
    dispatch(updateName({ key: playerKey, value: e.target.value }));
  };

  const handleKeyPress = e => {
    if (e.which === ENTER_KEY && e.target.value) {
      setIsEditing(false);
      dispatch(updateName({ key: playerKey, value: e.target.value }));
    }
  };

  const startEditing = () => {
    setIsEditing(true);
    setTimeout(() => {
      inputRef && inputRef.current && inputRef.current.focus();
    });
  };

  if (isEditing) {
    return (
      <input
        ref={inputRef}
        className={classes.editNameField}
        defaultValue={name}
        type="text"
        onBlur={handleBlur}
        onKeyPress={handleKeyPress}
      />
    )
  }

  return <em className={classes.staticName} onClick={startEditing}>{name}<span><EditIcon classes={classes} /></span></em>
};

const styles = {
  editNameField: {
    padding: "5px",
  },
  staticName: {
    color: "#ae3ec9",
    fontSize: "24px",
    fontWeight: "700",
    cursor: "pointer",
  },
  editIcon: {
    fill: "#e599f7",
    height: "14px",
    marginLeft: "10px",
  },
};

export const mapStateToProps = (state, ownProps) => ({
  name: state.game[ownProps.playerKey].name,
});

export default connect(mapStateToProps)(withStyles(styles)(PlayerField));