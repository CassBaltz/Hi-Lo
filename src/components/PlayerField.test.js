import React from "react";

import { PlayerField, styles, } from "./PlayerField";

const props = {
  classes: classReducer(styles),
  dispatch: jest.fn(),
  name: "mike",
  playerKey: "PLAYER_ONE",
};


describe("<PlayerField />", () => {
  it("should render correctly when the field is not being edited", () => {
    const playerField = shallow(<PlayerField {...props} />);

    expect(playerField.find("em.staticName").length).toBe(1);
    expect(playerField.find("input.editNameField").length).toBe(0);

    expect(toJson(playerField)).toMatchSnapshot();
  });

  it("should show the edit field on click and render the edit field correctly", () => {
    const playerField = shallow(<PlayerField {...props} />);

    playerField.find("em.staticName").simulate("click");
    expect(playerField.find("em.staticName").length).toBe(0);
    expect(playerField.find("input.editNameField").length).toBe(1);

    expect(toJson(playerField)).toMatchSnapshot();
  });
});