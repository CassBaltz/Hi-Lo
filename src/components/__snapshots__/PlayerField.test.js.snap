// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlayerField /> should render correctly when the field is not being edited 1`] = `
<em
  className="staticName"
  onClick={[Function]}
>
  mike
  <span>
    <EditIcon
      classes={
        Object {
          "editIcon": "editIcon",
          "editNameField": "editNameField",
          "staticName": "staticName",
        }
      }
    />
  </span>
</em>
`;

exports[`<PlayerField /> should show the edit field on click and render the edit field correctly 1`] = `
<input
  className="editNameField"
  defaultValue="mike"
  onBlur={[Function]}
  onKeyPress={[Function]}
  type="text"
/>
`;
